src/mptool.f90

```diff
@@ -143,5 +290,343 @@ pure function cross_product(a,b)
+      call mpi_file_write(fh, data_6, local_size, real_type, status, ierr)
+
+    endif
+
+    ! close the file
+    call mpi_file_close(fh, ierr)
+
+    if(nrank==0) print*,'<< ',filename
+
+  end subroutine pwrite_2darray
+
+  subroutine pread_1darray(filename,data2read)
+
+    ! arguments
+    character(len=*),intent(in) :: filename
+    real(mytype),intent(inout),allocatable :: data2read(:)
+
+    ! local data
+    integer :: local_size, ierr, fh, datatype, status(mpi_status_size),datasize
+    integer(kind=mpi_offset_kind) :: offset
+
+    ! calcualte local size
+    local_size=size(data2read)
+
+    ! calculate the offset for each process
+    offset = prelay(local_size)*mytype
+
+    ! Open the file in read mode using MPI-IO
+    call MPI_FILE_OPEN(MPI_COMM_WORLD, filename, MPI_MODE_RDONLY, MPI_INFO_NULL, fh, ierr)
+  
+    ! Set the file view for each process
+    call MPI_FILE_SET_VIEW(fh, offset, real_type, real_type, 'native', MPI_INFO_NULL, ierr)
+  
+    ! Read the local portion of the array from the file
+    call MPI_FILE_READ(fh, data2read, local_size, real_type, status, ierr)
+  
+    ! Close the file
+    call MPI_FILE_CLOSE(fh, ierr)
+
+    datasize=psum(local_size)
+
+    if(nrank==0) print*,'>> ',filename,' data size:',datasize
+
+
+  end subroutine pread_1darray
+
+
+  subroutine pread_2darray(filename,data2read)
+
+    ! arguments
+    character(len=*),intent(in) :: filename
+    real(mytype),intent(inout),allocatable :: data2read(:,:)
+
+    ! local data
+    integer :: local_size, ierr, fh, datatype, status(mpi_status_size),datasize
+    integer(kind=mpi_offset_kind) :: offset
+
+
+    ! calcualte local size
+    local_size=size(data2read)
+
+    ! calculate the offset for each process
+    offset = prelay(local_size)*mytype
+
+    ! Open the file in read mode using MPI-IO
+    call MPI_FILE_OPEN(MPI_COMM_WORLD, filename, MPI_MODE_RDONLY, MPI_INFO_NULL, fh, ierr)
+  
+    ! Set the file view for each process
+    call MPI_FILE_SET_VIEW(fh, offset, real_type, real_type, 'native', MPI_INFO_NULL, ierr)
+  
+    ! Read the local portion of the array from the file
+    call MPI_FILE_READ(fh, data2read, local_size, real_type, status, ierr)
+  
+    ! Close the file
+    call MPI_FILE_CLOSE(fh, ierr)
+
+    local_size=size(data2read,2)
+
+    datasize=psum(local_size)
+
+    if(nrank==0) print*,'>> ',filename,' data size:',datasize
+
+
+  end subroutine pread_2darray
+
+  ! this function return the number add from all processors before it
+  integer function prelay(number)
+
+    ! arguments
+    integer,intent(in) :: number
+
+    ! local data
+    integer :: table(nproc)
+    integer :: ierr,i
+
+    call mpi_allgather(number,1,mpi_integer,                              &
+                       table,1,mpi_integer,mpi_comm_world,ierr)
+
+    prelay=0
+    do i=1,nrank
+      prelay=prelay+table(i)
+    enddo
+    
+    return
+
+  end function prelay
   !
+  function linintp(xx1,xx2,yy1,yy2,xx) result(yy)
```

Comment:
Please add a comment describing the objective of the function and the limitations (`yy1` and `yy2` must be 3D arrays with the same size)

Reply:
I think this conversation has been resolved based on the following comment/conversation.